libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1361,26 +1394,143 @@ void IRGeneratorForStatements::appendAndOrOperatorCode(BinaryOperation const& _b
 
 	_binOp.leftExpression().accept(*this);
 
-	string value = m_context.variable(_binOp);
-	m_code << "let " << value << " := " << m_context.variable(_binOp.leftExpression()) << "\n";
+	IRVariable value(_binOp);
+	define(value, _binOp.leftExpression());
 	if (op == Token::Or)
-		m_code << "if iszero(" << value << ") {\n";
+		m_code << "if iszero(" << value.name() << ") {\n";
 	else
-		m_code << "if " << value << " {\n";
+		m_code << "if " << value.name() << " {\n";
 	_binOp.rightExpression().accept(*this);
-	m_code << value << " := " + m_context.variable(_binOp.rightExpression()) << "\n";
+	assign(value, _binOp.rightExpression());
 	m_code << "}\n";
 }
 
-void IRGeneratorForStatements::setLValue(Expression const& _expression, unique_ptr<IRLValue> _lvalue)
+void IRGeneratorForStatements::writeToLValue(IRLValue const& _lvalue, IRVariable const& _value)
+{
+	std::visit(
+		util::GenericVisitor{
+			[&](IRLValue::Storage const& _storage) {
+				std::optional<unsigned> offset;
+
+				if (std::holds_alternative<unsigned>(_storage.offset))
+					offset = std::get<unsigned>(_storage.offset);
+
+				IRVariable prepared(m_context.newYulVariable(), _lvalue.type);
```

Comment:
I think this should not convert - conversion is done in `updateStorageValueFunction`, at least for some types.